### DISCUSSION TITLE
How to add options

### DISCUSSION THREAD

**Discussion by c350147221** [seleniumbase/SeleniumBase#1221] [Q&A] [answered]
hi，

How to add options？

as follows

opt = Options()
opt.add_experimental_option("prefs", {"profile.default_content_setting_values.media_stream_mic": 1,
                                      "profile.default_content_setting_values.media_stream_camera": 1,
                                      "profile.default_content_setting_values.geolocation": 1,
                                      "profile.default_content_setting_values.notifications": 1})
driver = webdriver.Chrome("chromedriver", options=opt)

**Comment by mdmintz** [comment, marked as answer]
@c350147221 There's ``--chromium-arg=ARG`` for a comma-separated list of Chromium args, but that's for ``chrome_options.add_argument()``. For changing the profile, you'll need to override ``get_new_driver()`` from BaseCase. 
Here's an example of that:

```python
from selenium import webdriver
from seleniumbase import BaseCase


class OverrideDriverTest(BaseCase):

    def get_new_driver(self, *args, **kwargs):
        """ This method overrides get_new_driver() from BaseCase. """
        options = webdriver.ChromeOptions()
        options.add_experimental_option(
            "excludeSwitches", ["enable-automation"]
        )
        options.add_experimental_option("useAutomationExtension", False)
        if self.headless:
            options.add_argument("--headless")
        return webdriver.Chrome(options=options)

    def test_simple(self):
        self.open("https://seleniumbase.io/demo_page")
        self.assert_text("Demo Page", "h1")
```

**Reply by webmoskal** [reply]
can you provide an example when using SB as fixture?

**Reply by mdmintz** [reply]
@webmoskal https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_sb_fixture.py

**Reply by nakodavati** [reply]
i have used this to override it  def get_new_driver(self, *args, **kwargs):
        """ This method overrides get_new_driver() from BaseCase. """
        options = webdriver.ChromeOptions()
        options.add_experimental_option(
            "excludeSwitches", ["enable-automation"]
        )
        options.add_experimental_option("useAutomationExtension", False)
        if self.headless:
            options.add_argument("--headless")
        return webdriver.Chrome(options=options) 

but getting error when i use self.open("website")--> getting outofscope exception for self

note: i am creating one setUP (self) with self.get_new_driver(self) method call and samepage i have defined get_new_driver method

**Reply by mdmintz** [reply]
@nakodavati If overriding ``setUp()`` and/or ``tearDown()``, you need to call ``super()`` in them so that important setup steps in BaseCase ``setUp()`` and ``tearDown()`` methods aren't skipped. (See https://github.com/seleniumbase/SeleniumBase/blob/master/examples/boilerplates/base_test_case.py)

**Reply by nakodavati** [reply]
Inside setUP method ,if i use super().setUp() i am getting new automatic browser opening by seleniumbase , actually i want to avoid that -so trying to override get_new_driver method , and without "super().setUp()" if i use self.methodcall i am gettting Outofscope exception.how to solve this

**Reply by mdmintz** [reply]
Follow the model for overriding ``get_new_driver()`` here: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_override_driver.py
Follow the model for overriding ``setUp()`` and ``tearDown()`` here: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/boilerplates/base_test_case.py

**Reply by nakodavati** [reply]
@mdmintz 
I understand that by overriding get_new_driver() method,  I am able to add chrome options. 

When I add code as below:      

def setUp(self):
        super().setUp()
        self.get_new_driver()

Two chrome browsers are opened. One through super().setup() and the other through get_new_driver call.  I'm required to open only one browser with my chrome options. Is there any way to achieve this?

**Reply by mdmintz** [reply]
``self.get_new_driver()`` is already called by the ``super().setUp()`` part of the ``setUp()`` method. It should not be added directly to that custom ``setUp()`` method.

**Reply by hasanzaibak** [reply]
if i added get_new_driver for example with only headless, will this make the class run the seleniumbase default browser or selenium original one? I am looking of how can i be able to add options but at the same time, don't lost the advantages of seleniumbase for evading the antibot systems. Your help would be great with this @mdmintz

**Reply by mdmintz** [reply]
@hasanzaibak get_new_driver() launches a browser with SeleniumBase settings. For UC Mode, you’ll need to set undetectable=True.

**Reply by hasanzaibak** [reply]
@mdmintz Thank you for your reply. In the docs i've seen uc=true is added with Driver(), how can i use it when adding the options and in this example, with using         driver = webdriver.Chrome(options=options)?

**Reply by mdmintz** [reply]
@hasanzaibak It can’t be called via webdriver.Chrome(), as that is just raw Selenium without any SeleniumBase options or modes.

**Reply by hasanzaibak** [reply]
@mdmintz can i achieve the same uc results with launching webdriver.Chrome with driver.uc_open(url)?

**Reply by mdmintz** [reply]
@hasanzaibak driver.uc_open(url) is for opening a URL after the browser has already been launched.

**Reply by hasanzaibak** [reply]
@mdmintz I am sorry then I am lost, how will i be able to use or add options when needed, and at the same time use the uc mode? This is the main feature to be honest that I am trying seleniumbase for.

**Reply by mdmintz** [reply]
@hasanzaibak pass in the options you need: https://github.com/seleniumbase/SeleniumBase/blob/e87d7384298a70c1d308159b8c67766a90972447/seleniumbase/plugins/driver_manager.py#L65

**Reply by hasanzaibak** [reply]
@mdmintz I have my crawler modules ready with selenium and I am trying to refactor with seleniumbase. This seems to be working but I am unable to see the browser even with headless false, I am not getting any errors too. Any suggestions? 

`

class BrowserManager:
    def __init__(self):
        self.driver = self.setup_driver()
        self.logger = logging.getLogger(__name__)
        self.start_url = config.crawler_settings.start_url

    @staticmethod
    def setup_driver() -> Driver:
        driver = Driver(
            "chrome",
            undetectable=True,
            uc_subprocess=True,
            incognito=True,
            d_width=1920,
            d_height=1080,
            headless=False,
            headless2=False,
            
        )
        return driver

    def open_url(self) -> None:
        try:
            self.driver.uc_open(self.start_url)
            self.logger.info(f"Opened URL: {self.start_url}")
        except Exception as e:
            self.logger.error(f"Error opening URL: {str(e)}")
            raise

    def driver_quit(self) -> None:
        self.driver.quit()

`

**Reply by mdmintz** [reply]
Using Linux? That defaults to headless mode unless you add `headed=True`.

**Reply by hasanzaibak** [reply]
@mdmintz Yes I use linux. It worked greatly with headed thank you. But i was wondering, will i achieve the same results with this code or is it better to use BaseCase class? Will it have difference in the results? Can you suggest some docs or readings for this?

**Reply by mdmintz** [reply]
See [SeleniumBase Syntax Formats](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md). The `BaseCase` and `SB()` formats will have more abilities than the `Driver()` format.

**Comment by samliangsk** [comment]
If you are only trying to enable camera and audio, the answer in post #1566 would be easier.

**Comment by naresh237** [comment]
@mdmintz  i am using docker setup for my seleniumbase test cases, using  chrome version 111.0.5563.19-1, and chrome driver version 111.0.5563.19, python version 3.9.6 , seleniumbase==4.14.8 ,    selenium==4.9.1,  it seems there is no compatibility issues here , but when i am running inside my container i am getting " urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))" this error , tried all options and not solved yet . i have overrided get_new_driver and using this     def get_new_driver(self, *args, **kwargs):
        """ This method overrides get_new_driver() from BaseCase. """
        options = webdriver.ChromeOptions()
        #options.add_arguments("start-maximized")
        options.add_argument("--no-sandbox")
        options.add_argument("--headless")
        options.add_experimental_option("useAutomationExtension", False)
        options.add_argument("--disable-dev-shm-usage")
        options.add_argument("--remote-debugging-port=9222")
        options.add_argument("--ignore-certificate-errors")
        options.add_argument("--allow-running-insecure-content")
        options.add_argument("--ignore-urlfetcher-cert-requests")
        options.add_argument("--disable-blink-features=AutomationControlled")
        print(options)
        driver = webdriver.Chrome(executable_path="/usr/bin/chromedriver", options=options)
        return driver.  could you please help on this

**Reply by mdmintz** [reply]
@naresh237 Upgrade to the latest version of seleniumbase and use the default settings. Don't override ``get_new_driver()``. If still having issues after trying that, use this Dockerfile: https://github.com/seleniumbase/SeleniumBase/blob/master/Dockerfile

**Reply by naresh237** [reply]
I have updated my seleniumbase to 4.15.15 latest version . now i am getting error on pytest fixtures not found . 
i am having conftest.py file under testcases and i am initialising all page objects here 
@pytest.fixture()
def get_pages(request):
    login_page = LoginPage()

now i am calling my pytest fixture @pytest.mark.userfixture("get_pages").  after gone through latest version changes for seleniumbase , in my test method i am using def test_method(self,request) page = request.getfixturevalue('get_pages'). even in both the cases i am getting pytest fixture not found issue

**Reply by mdmintz** [reply]
See the example of tests that use ``getfixturevalue()``: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_request_sb_fixture.py
Run that, and once you've confirmed that it works, modify your tests to match that pattern/structure.

**Reply by naresh237** [reply]
@mdmintz  One more question , if i want to call a method(which is having fixture parameter -->here in my case it should be request fixture),  now how to call the method and what are the arguments we need to provide , i have tried as below and not working , is that possible ?

class Test_Request_Fixture:
    def test_request_sb_fixture_in_class(self, request):
        sb = request.getfixturevalue("sb")
        sb.open("https://seleniumbase.io/demo_page")
        sb.assert_element("input#myTextInput")
        sb.type("#myTextarea", "Automated")
        sb.assert_text("This Text is Green", "#pText")
        sb.click("#myButton")
        sb.assert_text("This Text is Purple", "#pText")
        sb.tearDown()

    def my_method(self):
        request = pytest.FixtureRequest
        self.test_request_sb_fixture_in_class(request)--> calling above method like this 

this is as example i have added but this is the way i want to do in one scenario

**Reply by mdmintz** [reply]
Tests with fixtures must be called via ``pytest``. They cannot be called via a separate method.

**Reply by naresh237** [reply]
@mdmintz  i have updated to latest seleniumbase version 4.15.15 and not overriding get new driver and using the same which is provided by seleniumbase , still getting the same error while executing my tests in container "urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))" 

i have used same docker file you have shared ( it is not working while downloading 'goole-chrome-stable' hence done few tweets and successfully built the image)-- but while running tests still the same error . 

Note: local execution is working fine

**Comment by mdmintz** [comment]
Looks like you have a typo in your modified version. Should be “google-chrome-stable”.

**Reply by nakodavati** [reply]
its just a typo here, as i already said that i have built the image successfully and i have proper chromedriver and google-chrome-stable versions
Google Chrome 114.0.5735.198 
ChromeDriver 114.0.5735.90

google-chrome-stable --version
Google Chrome 114.0.5735.198 

and seleniumbase version as below
seleniumbase==4.16.1
selenium==4.10.0;python_version>="3.7"
urllib3>=1.26.16,<2;python_version>="3.7" and python_version<"3.10"
requests==2.31.0;python_version>="3.7"
pytest==7.4.0;python_version>="3.7"

when executing tests getting below error
 urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

i am executing inside docker container as a non-root user

**Reply by mdmintz** [reply]
It may be related to https://stackoverflow.com/questions/50597923/unable-to-requests-get-a-website-remote-end-closed-connection-without-respon, but I can't know for sure unless you tell me what line of SeleniumBase code caused the issue. If it's related to ``requests.get()``, then https://stackoverflow.com/a/50598158/7058266 would have the solution.

**Reply by SMTamim** [reply]
How can it be done with 

with SB(headed=True, headless=False, block_images=True, uc=True) as sb:

**Reply by mdmintz** [reply]
`SB()` options are defined here:

https://github.com/seleniumbase/SeleniumBase/blob/570910c6f6e079cddd00c23e826e202c6489f39e/seleniumbase/plugins/sb_manager.py#L30

For ones that aren't explicitly defined, add them to `chromium_arg` (comma-separated with no spaces).

**Reply by SMTamim** [reply]
Thanks.
